Compiler crashes on inner class that extends sealed class

```scala
sealed class Foo

class Test {
  def f = {
    class Bar extends Foo
  }
}
```

<details>
  <summary>Stacktrace</summary>
  <pre>
exception occurred while compiling tests/allan/Test.scala
Exception in thread "main" java.lang.AssertionError: assertion failed: unresolved symbols: class Bar when pickling tests/allan/Test.scala
	at scala.Predef$.assert(Predef.scala:219)
	at dotty.tools.dotc.core.tasty.TreePickler.pickle(TreePickler.scala:592)
	at dotty.tools.dotc.transform.Pickler.$anonfun$run$3(Pickler.scala:52)
	at dotty.tools.dotc.transform.Pickler.$anonfun$run$3$adapted(Pickler.scala:45)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at dotty.tools.dotc.transform.Pickler.$anonfun$run$2(Pickler.scala:45)
	at dotty.tools.dotc.transform.Pickler.$anonfun$run$2$adapted(Pickler.scala:44)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at dotty.tools.dotc.transform.Pickler.run(Pickler.scala:44)
	at dotty.tools.dotc.core.Phases$Phase.$anonfun$runOn$1(Phases.scala:288)
	at scala.collection.immutable.List.map(List.scala:283)
	at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:286)
	at dotty.tools.dotc.core.Phases$Phase.runOn$(Phases.scala:285)
	at dotty.tools.dotc.transform.Pickler.runOn(Pickler.scala:76)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$3(Run.scala:124)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.trackTime(Stats.scala:47)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$2(Run.scala:122)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$2$adapted(Run.scala:120)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:32)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:29)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:191)
	at dotty.tools.dotc.Run.$anonfun$compileUnits$1(Run.scala:120)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:86)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:108)
	at dotty.tools.dotc.Run.compileSources(Run.scala:95)
	at dotty.tools.dotc.Run.compile(Run.scala:79)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:26)
	at dotty.tools.dotc.Driver.process(Driver.scala:124)
	at dotty.tools.dotc.Driver.process(Driver.scala:93)
	at dotty.tools.dotc.Driver.process(Driver.scala:105)
	at dotty.tools.dotc.Driver.main(Driver.scala:132)
	at dotty.tools.dotc.Main.main(Main.scala)
java.lang.RuntimeException: Nonzero exit code returned from runner: 1
	at scala.sys.package$.error(package.scala:27)
  </pre>
</details>
<br>

--- a/compiler/src/dotty/tools/dotc/core/Annotations.scala
+++ b/compiler/src/dotty/tools/dotc/core/Annotations.scala
@@ -142,15 +142,28 @@ object Annotations {
       apply(defn.AliasAnnot, List(
         ref(TermRef(sym.owner.thisType, sym.name, sym))))
 
-    def makeChild(delayedSym: Context => Symbol)(implicit ctx: Context): Annotation = {
-      def makeChildLater(implicit ctx: Context) = {
-        val sym = delayedSym(ctx)
-        New(defn.ChildAnnotType.appliedTo(sym.owner.thisType.select(sym.name, sym)), Nil)
-      }
-      deferred(defn.ChildAnnot, implicit ctx => makeChildLater(ctx))
-    }
+    /** Extractor for child annotations */
+    object Child {
 
-    def makeChild(sym: Symbol)(implicit ctx: Context): Annotation = makeChild(_ => sym)
+      /** A deferred annotation to the result of a given child computation */
+      def apply(delayedSym: Context => Symbol)(implicit ctx: Context): Annotation = {
+        def makeChildLater(implicit ctx: Context) = {
+          val sym = delayedSym(ctx)
+          New(defn.ChildAnnotType.appliedTo(sym.owner.thisType.select(sym.name, sym)), Nil)
+        }
+        deferred(defn.ChildAnnot, implicit ctx => makeChildLater(ctx))
+      }
+
+      /** A regular, non-deferred Child annotation */
+      def apply(sym: Symbol)(implicit ctx: Context): Annotation = apply(_ => sym)
+
+      def unapply(ann: Annotation)(implicit ctx: Context): Option[Symbol] =
+        if (ann.symbol == defn.ChildAnnot) {
+          val AppliedType(tycon, (arg: NamedType) :: Nil) = ann.tree.tpe
+          Some(arg.symbol)
+        }
+        else None
+    }
 
     def makeSourceFile(path: String)(implicit ctx: Context) =
       apply(defn.SourceFileAnnot, Literal(Constant(path)))

--- a/compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala
+++ b/compiler/src/dotty/tools/dotc/core/classfile/ClassfileParser.scala
@@ -258,7 +258,7 @@ class ClassfileParser(
             ctx.warning(s"no linked class for java enum $sym in ${sym.owner}. A referencing class file might be missing an InnerClasses entry.")
           else {
             if (!(enumClass is Flags.Sealed)) enumClass.setFlag(Flags.AbstractSealed)
-            enumClass.addAnnotation(Annotation.makeChild(sym))
+            enumClass.addAnnotation(Annotation.Child(sym))
           }
         }
       } finally {

--- a/compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala
+++ b/compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala
@@ -570,12 +570,11 @@ class TreePickler(pickler: TastyPickler) {
       if (flags is Covariant) writeByte(COVARIANT)
       if (flags is Contravariant) writeByte(CONTRAVARIANT)
     }
-    sym.annotations.foreach(pickleAnnotation)
+    sym.annotations.foreach(pickleAnnotation(sym, _))
   }
 
-  private def isUnpicklable(ann: Annotation)(implicit ctx: Context) = ann match {
-    case Annotation.Child(sym) =>
-      sym.isLocal && ctx.owner.topLevelClass != sym.topLevelClass
+  private def isUnpicklable(owner: Symbol, ann: Annotation)(implicit ctx: Context) = ann match {
+    case Annotation.Child(sym) => sym.isInaccessibleChildOf(owner)
       // If child annotation refers to a local class or enum value under
       // a different toplevel class, it is impossible to pickle a reference to it.
       // Such annotations will be reconstituted when unpickling the child class.
@@ -584,8 +583,8 @@ class TreePickler(pickler: TastyPickler) {
       // inline bodies are reconstituted automatically when unpickling
   }
 
-  def pickleAnnotation(ann: Annotation)(implicit ctx: Context) =
-    if (!isUnpicklable(ann)) {
+  def pickleAnnotation(owner: Symbol, ann: Annotation)(implicit ctx: Context) =
+    if (!isUnpicklable(owner, ann)) {
       writeByte(ANNOTATION)
       withLength { pickleType(ann.symbol.typeRef); pickleTree(ann.tree) }
     }

--- a/compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala
+++ b/compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala
@@ -573,8 +573,19 @@ class TreePickler(pickler: TastyPickler) {
     sym.annotations.foreach(pickleAnnotation)
   }
 
+  private def isUnpicklable(ann: Annotation)(implicit ctx: Context) = ann match {
+    case Annotation.Child(sym) =>
+      sym.isLocal && ctx.owner.topLevelClass != sym.topLevelClass
+      // If child annotation refers to a local class or enum value under
+      // a different toplevel class, it is impossible to pickle a reference to it.
+      // Such annotations will be reconstituted when unpickling the child class.
+      // See tests/pickling/i3149.scala
+    case _ => ann.symbol == defn.BodyAnnot
+      // inline bodies are reconstituted automatically when unpickling
+  }
+
   def pickleAnnotation(ann: Annotation)(implicit ctx: Context) =
-    if (ann.symbol != defn.BodyAnnot) { // inline bodies are reconstituted automatically when unpickling
+    if (!isUnpicklable(ann)) {
       writeByte(ANNOTATION)
       withLength { pickleType(ann.symbol.typeRef); pickleTree(ann.tree) }
     }

--- a/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
+++ b/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
@@ -705,7 +705,7 @@ class TreeUnpickler(reader: TastyReader, nameAtRef: NameRef => TermName, posUnpi
       if ((sym.isClass || sym.is(CaseVal)) && sym.isLocal)
         // Child annotations for local classes and enum values are not pickled, so
         // need to be re-established here.
-        sym.registerIfChild()
+        sym.registerIfChild(late = true)
       tree
     }
 

--- a/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
+++ b/compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala
@@ -11,6 +11,7 @@ import util.Positions._
 import ast.{tpd, Trees, untpd}
 import Trees._
 import Decorators._
+import transform.SymUtils._
 import TastyUnpickler._, TastyBuffer._
 import scala.annotation.{tailrec, switch}
 import scala.collection.mutable.ListBuffer
@@ -663,7 +664,6 @@ class TreeUnpickler(reader: TastyReader, nameAtRef: NameRef => TermName, posUnpi
             def isCodefined =
               roots.contains(companion.denot) == seenRoots.contains(companion)
             if (companion.exists && isCodefined) {
-              import transform.SymUtils._
               if (sym is Flags.ModuleClass) sym.registerCompanionMethod(nme.COMPANION_CLASS_METHOD, companion)
               else sym.registerCompanionMethod(nme.COMPANION_MODULE_METHOD, companion)
             }
@@ -702,6 +702,10 @@ class TreeUnpickler(reader: TastyReader, nameAtRef: NameRef => TermName, posUnpi
       if (!sym.isType) { // Only terms might have leaky aliases, see the documentation of `checkNoPrivateLeaks`
         sym.info = ta.avoidPrivateLeaks(sym, tree.pos)
       }
+      if ((sym.isClass || sym.is(CaseVal)) && sym.isLocal)
+        // Child annotations for local classes and enum values are not pickled, so
+        // need to be re-established here.
+        sym.registerIfChild()
       tree
     }
 

--- a/compiler/src/dotty/tools/dotc/core/unpickleScala2/Scala2Unpickler.scala
+++ b/compiler/src/dotty/tools/dotc/core/unpickleScala2/Scala2Unpickler.scala
@@ -838,7 +838,7 @@ class Scala2Unpickler(bytes: Array[Byte], classRoot: ClassDenotation, moduleClas
       val start = readIndex
       readNat() // skip reference for now
       target.addAnnotation(
-          Annotation.makeChild(implicit ctx =>
+          Annotation.Child(implicit ctx =>
               atReadPos(start, () => readSymbolRef())))
     }
   }

--- a/compiler/src/dotty/tools/dotc/transform/PostTyper.scala
+++ b/compiler/src/dotty/tools/dotc/transform/PostTyper.scala
@@ -14,7 +14,7 @@ import SymDenotations._, Symbols._, StdNames._, Annotations._, Trees._, Scopes._
 import util.Positions._
 import Decorators._
 import config.Printers.typr
-import Symbols._, TypeUtils._
+import Symbols._, TypeUtils._, SymUtils._
 import reporting.diagnostic.messages.SuperCallsNotAllowedInline
 
 /** A macro transform that runs immediately after typer and that performs the following functions:
@@ -124,14 +124,9 @@ class PostTyper extends MacroTransform with SymTransformer  { thisTransformer =>
     private def transformAnnot(annot: Annotation)(implicit ctx: Context): Annotation =
       annot.derivedAnnotation(transformAnnot(annot.tree))
 
-    private def registerChild(sym: Symbol, tp: Type)(implicit ctx: Context) = {
-      val cls = tp.classSymbol
-      if (cls.is(Sealed)) cls.addAnnotation(Annotation.makeChild(sym))
-    }
-
     private def transformMemberDef(tree: MemberDef)(implicit ctx: Context): Unit = {
       val sym = tree.symbol
-      if (sym.is(CaseVal, butNot = Method | Module)) registerChild(sym, sym.info)
+      sym.registerIfChild()
       sym.transformAnnotations(transformAnnot)
     }
 
@@ -259,12 +254,7 @@ class PostTyper extends MacroTransform with SymTransformer  { thisTransformer =>
               sym.addAnnotation(Annotation.makeSourceFile(ctx.compilationUnit.source.file.path))
 
             // Add Child annotation to sealed parents unless current class is anonymous
-            if (!sym.isAnonymousClass) // ignore anonymous class
-              sym.asClass.classParents.foreach { parent =>
-                val sym2 = if (sym.is(Module)) sym.sourceModule else sym
-                registerChild(sym2, parent)
-              }
-
+            sym.registerIfChild()
             tree
           }
           super.transform(tree)

--- a/compiler/src/dotty/tools/dotc/transform/PostTyper.scala
+++ b/compiler/src/dotty/tools/dotc/transform/PostTyper.scala
@@ -252,9 +252,6 @@ class PostTyper extends MacroTransform with SymTransformer  { thisTransformer =>
               ctx.compilationUnit.source.exists &&
               sym != defn.SourceFileAnnot)
               sym.addAnnotation(Annotation.makeSourceFile(ctx.compilationUnit.source.file.path))
-
-            // Add Child annotation to sealed parents unless current class is anonymous
-            sym.registerIfChild()
             tree
           }
           super.transform(tree)

--- a/compiler/src/dotty/tools/dotc/transform/SymUtils.scala
+++ b/compiler/src/dotty/tools/dotc/transform/SymUtils.scala
@@ -137,11 +137,13 @@ class SymUtils(val self: Symbol) extends AnyVal {
 
   /** If this symbol is an enum value or a named class, register it as a child
    *  in all direct parent classes which are sealed.
+   *   @param  @late  If true, register only inaccessible children (all others are already
+   *                  entered at this point).
    */
-  def registerIfChild()(implicit ctx: Context): Unit = {
+  def registerIfChild(late: Boolean = false)(implicit ctx: Context): Unit = {
     def register(child: Symbol, parent: Type) = {
       val cls = parent.classSymbol
-      if (cls.is(Sealed) && !cls.children.contains(child))
+      if (cls.is(Sealed) && (!late || child.isInaccessibleChildOf(cls)))
         cls.addAnnotation(Annotation.Child(child))
     }
     if (self.isClass && !self.isAnonymousClass)
@@ -152,6 +154,13 @@ class SymUtils(val self: Symbol) extends AnyVal {
     else if (self.is(CaseVal, butNot = Method | Module))
       register(self, self.info)
   }
+
+  /** Is this symbol defined locally (i.e. at some level owned by a term) and
+   *  defined in a different toplevel class than its supposed parent class `cls`?
+   *  Such children are not pickled, and have to be reconstituted manually.
+   */
+  def isInaccessibleChildOf(cls: Symbol)(implicit ctx: Context) =
+    self.isLocal && !cls.topLevelClass.isLinkedWith(self.topLevelClass)
 
   /** If this is a sealed class, its known children */
   def children(implicit ctx: Context): List[Symbol] =

--- a/compiler/src/dotty/tools/dotc/transform/SymUtils.scala
+++ b/compiler/src/dotty/tools/dotc/transform/SymUtils.scala
@@ -134,4 +134,36 @@ class SymUtils(val self: Symbol) extends AnyVal {
       }
     }
   }
+
+  /** If this symbol is an enum value or a named class, register it as a child
+   *  in all direct parent classes which are sealed.
+   */
+  def registerIfChild()(implicit ctx: Context): Unit = {
+    def register(child: Symbol, parent: Type) = {
+      val cls = parent.classSymbol
+      if (cls.is(Sealed) && !cls.children.contains(child))
+        cls.addAnnotation(Annotation.Child(child))
+    }
+    if (self.isClass && !self.isAnonymousClass)
+      self.asClass.classParents.foreach { parent =>
+        val child = if (self.is(Module)) self.sourceModule else self
+        register(child, parent)
+      }
+    else if (self.is(CaseVal, butNot = Method | Module))
+      register(self, self.info)
+  }
+
+  /** If this is a sealed class, its known children */
+  def children(implicit ctx: Context): List[Symbol] =
+    self.annotations.collect {
+      case Annotation.Child(child) => child
+    }
+
+  /** Is symbol directly or indirectly owned by a term symbol? */
+  @tailrec def isLocal(implicit ctx: Context): Boolean = {
+    val owner = self.owner
+    if (owner.isTerm) true
+    else if (owner.is(Package)) false
+    else owner.isLocal
+  }
 }

--- a/compiler/src/dotty/tools/dotc/transform/patmat/Space.scala
+++ b/compiler/src/dotty/tools/dotc/transform/patmat/Space.scala
@@ -514,12 +514,7 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
 
   /** Decompose a type into subspaces -- assume the type can be decomposed */
   def decompose(tp: Type): List[Space] = {
-    val children = tp.classSymbol.annotations.filter(_.symbol == ctx.definitions.ChildAnnot).map { annot =>
-      // refer to definition of Annotation.makeChild
-      annot.tree match {
-        case Apply(TypeApply(_, List(tpTree)), _) => tpTree.symbol
-      }
-    }
+    val children = tp.classSymbol.children
 
     debug.println(s"candidates for ${tp.show} : [${children.map(_.show).mkString(", ")}]")
 

--- a/compiler/src/dotty/tools/dotc/typer/Namer.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Namer.scala
@@ -470,7 +470,7 @@ class Namer { typer: Typer =>
     case vdef: ValDef if (isEnumConstant(vdef)) =>
       val enumClass = sym.owner.linkedClass
       if (!(enumClass is Flags.Sealed)) enumClass.setFlag(Flags.AbstractSealed)
-      enumClass.addAnnotation(Annotation.makeChild(sym))
+      enumClass.addAnnotation(Annotation.Child(sym))
     case _ =>
   }
 

--- a/tests/pickling/i3149.scala
+++ b/tests/pickling/i3149.scala
@@ -1,0 +1,7 @@
+sealed class Foo
+
+class Test {
+  def f = {
+    class Bar extends Foo
+  }
+}

--- a/tests/pos/i3149.scala
+++ b/tests/pos/i3149.scala
@@ -1,0 +1,14 @@
+sealed class Foo
+
+class Test {
+  def f = {
+    class Bar extends Foo
+  }
+  class C {
+    class Bar extends Foo
+  }
+  object O {
+    class Bar extends Foo
+  }
+}
+

--- a/tests/pos/i3149a.scala
+++ b/tests/pos/i3149a.scala
@@ -1,0 +1,14 @@
+sealed class Foo
+
+object Foo {
+  def f = {
+    class Bar extends Foo
+  }
+  class C {
+    class Bar extends Foo
+  }
+  object O {
+    class Bar extends Foo
+  }
+}
+

